File: 2025/jimmycho/week3/2025-06-22-Asteroid-Collision.py
class Solution:
    def asteroidCollision(self, asteroids: list[int]) -> list[int]:
        stack = [asteroids[0]]
        for current_asteroid in asteroids[1:]:
            while stack and current_asteroid < 0 and stack[-1] > 0:
                if abs(current_asteroid) > abs(stack[-1]):
                    stack.pop()
                elif abs(current_asteroid) == abs(stack[-1]):
                    stack.pop()
                    current_asteroid = 0
                    break
                else:
                    current_asteroid = 0
                    break
            if current_asteroid != 0:
                stack.append(current_asteroid)
        return stack

File: 2025/SoyoungJung/week3/2025-06-22-Asteroid-Collision.ts
function asteroidCollision(asteroids: number[]): number[] {
  const stack: number[] = [];

  for (const asteroid of asteroids) {
    let alive = true;

    while (
      alive &&
      asteroid < 0 &&
      stack.length > 0 &&
      stack[stack.length - 1] > 0
    ) {
      const top = stack[stack.length - 1];

      if (Math.abs(top) < Math.abs(asteroid)) {
        stack.pop();
      } else if (Math.abs(top) === Math.abs(asteroid)) {
        stack.pop();
        alive = false;
      } else {
        alive = false;
      }
    }

    if (alive) {
      stack.push(asteroid);
    }
  }

  return stack;
}
